### PR TITLE
[PM-13362] Add private key regeneration endpoint

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-13362

## 📔 Objective

The purpose of this PR is to add an endpoint to enable private key regeneration for the clients on login and unlock.
This is feature will be behind a feature flag.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: test/Api.Test/KeyManagement/Controllers/AccountsKeyManagementControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        await sutProvider.Sut.RegenerateKeysAsync(data);
+
+        await sutProvider.GetDependency<IOrganizationUserRepository>().Received(1)
+            .GetManyByUserAsync(Arg.Is(user.Id));
+        await sutProvider.GetDependency<IEmergencyAccessRepository>().Received(1)
+            .GetManyDetailsByGranteeIdAsync(Arg.Is(user.Id));
+        await sutProvider.GetDependency<IRegenerateUserAsymmetricKeysCommand>().ReceivedWithAnyArgs(1)
+            .RegenerateKeysAsync(Arg.Any<UserAsymmetricKeys>(),
+                Arg.Any<ICollection<OrganizationUser>>(),
+                Arg.Any<ICollection<EmergencyAccessDetails>>());
```
Comment (mzieniukbw) [inline test/Api.Test/KeyManagement/Controllers/AccountsKeyManagementControllerTests.cs:87]: We should check whether the data passed to the method is as expected

File: test/Core.Test/KeyManagement/Commands/RegenerateUserAsymmetricKeysCommandTests.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            .Received(1)
+            .RegenerateUserAsymmetricKeysAsync(Arg.Is(userAsymmetricKeys));
+    }
+
+    [Theory]
+    [BitAutoData(false, false, true)]
+    [BitAutoData(false, true, false)]
+    [BitAutoData(false, true, true)]
+    [BitAutoData(true, false, false)]
+    [BitAutoData(true, true, false)]
```
Comment (mzieniukbw) [inline test/Core.Test/KeyManagement/Commands/RegenerateUserAsymmetricKeysCommandTests.cs:59]: How about `[BitAutoData(true, false, true)]`

**Review by mzieniukbw — CHANGES_REQUESTED** [review]

File: test/Core.Test/KeyManagement/Commands/RegenerateUserAsymmetricKeysCommandTests.cs (hunk 2 above)
Comment (Thomas-Avery) [reply test/Core.Test/KeyManagement/Commands/RegenerateUserAsymmetricKeysCommandTests.cs:60]: Nice catch updated in https://github.com/bitwarden/server/pull/4929/commits/23d9e3f4e581874353c65bc587cb87806312ae51

File: test/Api.Test/KeyManagement/Controllers/AccountsKeyManagementControllerTests.cs (hunk 1 above)
Comment (Thomas-Avery) [reply test/Api.Test/KeyManagement/Controllers/AccountsKeyManagementControllerTests.cs]: Updated in https://github.com/bitwarden/server/pull/4929/commits/23d9e3f4e581874353c65bc587cb87806312ae51

File: test/Api.IntegrationTest/KeyManagement/Controllers/AccountsKeyManagementControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        {
+            await OrganizationTestHelpers.SignUpAsync(_factory,
+                PlanType.EnterpriseAnnually, _ownerEmail, passwordManagerSeats: 10,
+                paymentMethod: PaymentMethodType.Card);
+        }
+
+        if (hasDesignatedEmergencyAccess)
+        {
+            await CreateDesignatedEmergencyAccessAsync();
+        }
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply test/Api.IntegrationTest/KeyManagement/Controllers/AccountsKeyManagementControllerTests.cs]: Updated in https://github.com/bitwarden/server/pull/4929/commits/23d9e3f4e581874353c65bc587cb87806312ae51

File: src/Api/KeyManagement/Models/Requests/KeyRegenerationRequestModel.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+using System.ComponentModel.DataAnnotations;
+using Bit.Core.KeyManagement.Models.Data;
+using Bit.Core.Utilities;
+
+namespace Bit.Api.KeyManagement.Models.Requests;
+
+public class KeyRegenerationRequestModel
+{
+    [Required]
+    public required string UserPublicKey { get; set; }
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply src/Api/KeyManagement/Models/Requests/KeyRegenerationRequestModel.cs]: Updated in https://github.com/bitwarden/server/pull/4929/commits/4d6ae5741fdea0a5ebbf78306e78940eb5f30384

**Review by mzieniukbw — DISMISSED** [review]

**Review by mzieniukbw — DISMISSED** [review]

**Review by mzieniukbw — APPROVED** [review]

**Pull request merged**